File: src/org/thoughtcrime/securesms/components/ConversationItemThumbnail.java
package org.thoughtcrime.securesms.components;

import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.graphics.Path;
import android.graphics.RectF;
import android.support.annotation.ColorInt;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.annotation.UiThread;
import android.util.AttributeSet;
import android.widget.FrameLayout;
import android.widget.ImageView;

import org.thoughtcrime.securesms.R;
import org.thoughtcrime.securesms.attachments.Attachment;
import org.thoughtcrime.securesms.mms.GlideRequests;
import org.thoughtcrime.securesms.mms.Slide;
import org.thoughtcrime.securesms.mms.SlideClickListener;
import org.thoughtcrime.securesms.mms.SlidesClickedListener;
import org.thoughtcrime.securesms.util.ThemeUtil;

import java.util.List;

public class ConversationItemThumbnail extends FrameLayout {

  private static final String TAG = ConversationItemThumbnail.class.getSimpleName();

  private final float[] radii      = new float[8];
  private final RectF   bounds     = new RectF();
  private final Path    corners    = new Path();

  private ThumbnailView          thumbnail;
  private AlbumThumbnailView     album;
  private ImageView              shade;
  private ConversationItemFooter footer;
  private CornerMask             cornerMask;

  private final Paint outlinePaint = new Paint();
  {
    outlinePaint.setStyle(Paint.Style.STROKE);
    outlinePaint.setStrokeWidth(1f);
    outlinePaint.setAntiAlias(true);
  }

  public ConversationItemThumbnail(Context context) {
    super(context);
    init(null);
  }

  public ConversationItemThumbnail(Context context, AttributeSet attrs) {
    super(context, attrs);
    init(attrs);
  }

  public ConversationItemThumbnail(final Context context, AttributeSet attrs, int defStyle) {
    super(context, attrs, defStyle);
    init(attrs);
  }

  private void init(@Nullable AttributeSet attrs) {
    inflate(getContext(), R.layout.conversation_item_thumbnail, this);

    outlinePaint.setColor(ThemeUtil.getThemedColor(getContext(), R.attr.conversation_item_image_outline_color));

    this.thumbnail  = findViewById(R.id.conversation_thumbnail_image);
    this.album      = findViewById(R.id.conversation_thumbnail_album);
    this.shade      = findViewById(R.id.conversation_thumbnail_shade);
    this.footer     = findViewById(R.id.conversation_thumbnail_footer);
    this.cornerMask = new CornerMask(this);

    if (attrs != null) {
      TypedArray typedArray = getContext().getTheme().obtainStyledAttributes(attrs, R.styleable.ConversationItemThumbnail, 0, 0);
      thumbnail.setBounds(typedArray.getDimensionPixelSize(R.styleable.ConversationItemThumbnail_conversationThumbnail_minWidth, 0),
                          typedArray.getDimensionPixelSize(R.styleable.ConversationItemThumbnail_conversationThumbnail_maxWidth, 0),
                          typedArray.getDimensionPixelSize(R.styleable.ConversationItemThumbnail_conversationThumbnail_minHeight, 0),
                          typedArray.getDimensionPixelSize(R.styleable.ConversationItemThumbnail_conversationThumbnail_maxHeight, 0));
      typedArray.recycle();
    }
  }

  @SuppressWarnings("SuspiciousNameCombination")
  @Override
  protected void dispatchDraw(Canvas canvas) {
    if (cornerMask.isLegacy()) {
      cornerMask.mask(canvas);
    }

    super.dispatchDraw(canvas);

    if (!cornerMask.isLegacy()) {
      cornerMask.mask(canvas);
    }

    if (album.getVisibility() != VISIBLE) {
      final float halfStrokeWidth = outlinePaint.getStrokeWidth() / 2;

      bounds.left   = halfStrokeWidth;
      bounds.top    = halfStrokeWidth;
      bounds.right  = canvas.getWidth() - halfStrokeWidth;
      bounds.bottom = canvas.getHeight() - halfStrokeWidth;

      corners.reset();
      corners.addRoundRect(bounds, radii, Path.Direction.CW);

      canvas.drawPath(corners, outlinePaint);
    }
  }

  @Override
  public void setFocusable(boolean focusable) {
    thumbnail.setFocusable(focusable);
    album.setFocusable(focusable);
  }

  @Override
  public void setClickable(boolean clickable) {
    thumbnail.setClickable(clickable);
    album.setClickable(clickable);
  }

  @Override
  public void setOnLongClickListener(@Nullable OnLongClickListener l) {
    thumbnail.setOnLongClickListener(l);
    album.setOnLongClickListener(l);
  }

  public void showShade(boolean show) {
    shade.setVisibility(show ? VISIBLE : GONE);
    forceLayout();
  }

  public void setOutlineCorners(int topLeft, int topRight, int bottomRight, int bottomLeft) {
    radii[0] = radii[1] = topLeft;
    radii[2] = radii[3] = topRight;
    radii[4] = radii[5] = bottomRight;
    radii[6] = radii[7] = bottomLeft;

    cornerMask.setRadii(topLeft, topRight, bottomRight, bottomLeft);
  }

  public ConversationItemFooter getFooter() {
    return footer;
  }

  @UiThread
  public void setImageResource(@NonNull GlideRequests glideRequests, @NonNull List<Slide> slides,
                               boolean showControls, boolean isPreview)
  {
    if (slides.size() == 1) {
      thumbnail.setVisibility(VISIBLE);
      album.setVisibility(GONE);

      Attachment attachment = slides.get(0).asAttachment();
      thumbnail.setImageResource(glideRequests, slides.get(0), showControls, isPreview, attachment.getWidth(), attachment.getHeight());
      setTouchDelegate(thumbnail.getTouchDelegate());
    } else {
      thumbnail.setVisibility(GONE);
      album.setVisibility(VISIBLE);

      album.setSlides(glideRequests, slides, showControls);
      setTouchDelegate(album.getTouchDelegate());
    }
  }

  public void setConversationColor(@ColorInt int color) {
    if (album.getVisibility() == VISIBLE) {
      album.setCellBackgroundColor(color);
    }
  }

  public void setThumbnailClickListener(SlideClickListener listener) {
    thumbnail.setThumbnailClickListener(listener);
    album.setThumbnailClickListener(listener);
  }

  public void setDownloadClickListener(SlidesClickedListener listener) {
    thumbnail.setDownloadClickListener(listener);
    album.setDownloadClickListener(listener);
  }
}

File: app/src/main/java/org/thoughtcrime/securesms/components/ConversationItemThumbnail.kt
package org.thoughtcrime.securesms.components

import android.content.Context
import android.graphics.Canvas
import android.os.Bundle
import android.os.Parcelable
import android.util.AttributeSet
import android.widget.FrameLayout
import android.widget.ImageView
import androidx.annotation.ColorInt
import androidx.annotation.Px
import androidx.annotation.UiThread
import androidx.core.os.bundleOf
import org.signal.core.util.dp
import org.signal.core.util.getParcelableCompat
import org.thoughtcrime.securesms.R
import org.thoughtcrime.securesms.mms.GlideRequests
import org.thoughtcrime.securesms.mms.Slide
import org.thoughtcrime.securesms.mms.SlideClickListener
import org.thoughtcrime.securesms.mms.SlidesClickedListener
import org.thoughtcrime.securesms.util.Projection.Corners
import org.thoughtcrime.securesms.util.views.Stub

class ConversationItemThumbnail @JvmOverloads constructor(
  context: Context,
  attrs: AttributeSet? = null
) : FrameLayout(context, attrs) {

  private var state: ConversationItemThumbnailState
  private var thumbnail: Stub<ThumbnailView>
  private var album: Stub<AlbumThumbnailView>
  private var shade: ImageView
  var footer: Stub<ConversationItemFooter>
    private set
  private var cornerMask: CornerMask
  private var borderless = false
  private var normalBounds: IntArray
  private var gifBounds: IntArray
  private var minimumThumbnailWidth = 0
  private var maximumThumbnailHeight = 0

  init {
    inflate(context, R.layout.conversation_item_thumbnail, this)

    thumbnail = Stub(findViewById(R.id.thumbnail_view_stub))
    album = Stub(findViewById(R.id.album_view_stub))
    shade = findViewById(R.id.conversation_thumbnail_shade)
    footer = Stub(findViewById(R.id.footer_view_stub))
    cornerMask = CornerMask(this)

    var gifWidth = 260.dp

    if (attrs != null) {
      val typedArray = context.theme.obtainStyledAttributes(attrs, R.styleable.ConversationItemThumbnail, 0, 0)
      normalBounds = intArrayOf(
        typedArray.getDimensionPixelSize(R.styleable.ConversationItemThumbnail_conversationThumbnail_minWidth, 0),
        typedArray.getDimensionPixelSize(R.styleable.ConversationItemThumbnail_conversationThumbnail_maxWidth, 0),
        typedArray.getDimensionPixelSize(R.styleable.ConversationItemThumbnail_conversationThumbnail_minHeight, 0),
        typedArray.getDimensionPixelSize(R.styleable.ConversationItemThumbnail_conversationThumbnail_maxHeight, 0)
      )

      gifWidth = typedArray.getDimensionPixelSize(R.styleable.ConversationItemThumbnail_conversationThumbnail_gifWidth, gifWidth)

      typedArray.recycle()
    } else {
      normalBounds = intArrayOf(0, 0, 0, 0)
    }

    gifBounds = intArrayOf(
      gifWidth,
      gifWidth,
      1,
      Int.MAX_VALUE
    )

    minimumThumbnailWidth = -1
    maximumThumbnailHeight = -1

    state = ConversationItemThumbnailState()
  }

  override fun dispatchDraw(canvas: Canvas) {
    super.dispatchDraw(canvas)
    if (!borderless) {
      cornerMask.mask(canvas)
    }
  }

  override fun onSaveInstanceState(): Parcelable? {
    val root = super.onSaveInstanceState()
    return bundleOf(
      STATE_ROOT to root,
      STATE_STATE to state
    )
  }

  override fun onRestoreInstanceState(state: Parcelable) {
    if (state is Bundle && state.containsKey(STATE_STATE)) {
      val root: Parcelable? = state.getParcelableCompat(STATE_ROOT, Parcelable::class.java)
      this.state = state.getParcelableCompat(STATE_STATE, ConversationItemThumbnailState::class.java)!!
      super.onRestoreInstanceState(root)
    } else {
      super.onRestoreInstanceState(state)
    }
  }

  override fun setFocusable(focusable: Boolean) {
    state = state.copy(
      thumbnailViewState = state.thumbnailViewState.copy(focusable = focusable),
      albumViewState = state.albumViewState.copy(focusable = focusable)
    )

    state.applyState(thumbnail, album)
  }

  override fun setClickable(clickable: Boolean) {
    state = state.copy(
      thumbnailViewState = state.thumbnailViewState.copy(clickable = clickable),
      albumViewState = state.albumViewState.copy(clickable = clickable)
    )

    state.applyState(thumbnail, album)
  }

  override fun setOnLongClickListener(l: OnLongClickListener?) {
    state = state.copy(
      thumbnailViewState = state.thumbnailViewState.copy(longClickListener = l),
      albumViewState = state.albumViewState.copy(longClickListener = l)
    )

    state.applyState(thumbnail, album)
  }

  fun hideThumbnailView() {
    state = state.copy(thumbnailViewState = state.thumbnailViewState.copy(alpha = 0f))
    state.thumbnailViewState.applyState(thumbnail)
  }

  fun showThumbnailView() {
    state = state.copy(thumbnailViewState = state.thumbnailViewState.copy(alpha = 1f))
    state.thumbnailViewState.applyState(thumbnail)
  }

  val corners: Corners
    get() = Corners(cornerMask.radii)

  fun showShade(show: Boolean) {
    shade.visibility = if (show) VISIBLE else GONE
    forceLayout()
  }

  fun setCorners(topLeft: Int, topRight: Int, bottomRight: Int, bottomLeft: Int) {
    cornerMask.setRadii(topLeft, topRight, bottomRight, bottomLeft)
    state = state.copy(
      thumbnailViewState = state.thumbnailViewState.copy(
        cornerTopLeft = topLeft,
        cornerTopRight = topRight,
        cornerBottomRight = bottomRight,
        cornerBottomLeft = bottomLeft
      ),
      albumViewState = state.albumViewState.copy(
        cornerTopLeft = topLeft,
        cornerTopRight = topRight,
        cornerBottomRight = bottomRight,
        cornerBottomLeft = bottomLeft
      )
    )

    state.applyState(thumbnail, album)
  }

  fun setMinimumThumbnailWidth(@Px width: Int) {
    minimumThumbnailWidth = width
    state = state.copy(thumbnailViewState = state.thumbnailViewState.copy(minWidth = width))
    state.thumbnailViewState.applyState(thumbnail)
  }

  fun setMaximumThumbnailHeight(@Px height: Int) {
    maximumThumbnailHeight = height
    state = state.copy(thumbnailViewState = state.thumbnailViewState.copy(maxHeight = height))
    state.thumbnailViewState.applyState(thumbnail)
  }

  fun setBorderless(borderless: Boolean) {
    this.borderless = borderless
  }

  @UiThread
  fun setImageResource(
    glideRequests: GlideRequests,
    slides: List<Slide>,
    showControls: Boolean,
    isPreview: Boolean
  ) {
    if (slides.size == 1) {
      val slide = slides[0]

      if (slide.isVideoGif) {
        setThumbnailBounds(gifBounds)
      } else {
        setThumbnailBounds(normalBounds)

        if (minimumThumbnailWidth != -1) {
          state = state.copy(thumbnailViewState = state.thumbnailViewState.copy(minWidth = minimumThumbnailWidth))
        }

        if (maximumThumbnailHeight != -1) {
          state = state.copy(thumbnailViewState = state.thumbnailViewState.copy(maxHeight = maximumThumbnailHeight))
        }
      }

      state = state.copy(
        thumbnailViewState = state.thumbnailViewState.copy(visibility = VISIBLE),
        albumViewState = state.albumViewState.copy(visibility = GONE)
      )

      state.applyState(thumbnail, album)

      val attachment = slides[0].asAttachment()

      thumbnail.get().setImageResource(glideRequests, slides[0], showControls, isPreview, attachment.width, attachment.height)
      touchDelegate = thumbnail.get().touchDelegate
    } else {
      state = state.copy(
        thumbnailViewState = state.thumbnailViewState.copy(visibility = GONE),
        albumViewState = state.albumViewState.copy(visibility = VISIBLE)
      )

      state.applyState(thumbnail, album)
      album.get().setSlides(glideRequests, slides, showControls)
      touchDelegate = album.get().touchDelegate
    }
  }

  fun setConversationColor(@ColorInt color: Int) {
    state = state.copy(albumViewState = state.albumViewState.copy(cellBackgroundColor = color))
    state.albumViewState.applyState(album)
  }

  fun setThumbnailClickListener(listener: SlideClickListener?) {
    state = state.copy(
      thumbnailViewState = state.thumbnailViewState.copy(clickListener = listener),
      albumViewState = state.albumViewState.copy(clickListener = listener)
    )

    state.applyState(thumbnail, album)
  }

  fun setDownloadClickListener(listener: SlidesClickedListener?) {
    state = state.copy(
      thumbnailViewState = state.thumbnailViewState.copy(downloadClickListener = listener),
      albumViewState = state.albumViewState.copy(downloadClickListener = listener)
    )

    state.applyState(thumbnail, album)
  }

  private fun setThumbnailBounds(bounds: IntArray) {
    val (minWidth, maxWidth, minHeight, maxHeight) = bounds
    state = state.copy(
      thumbnailViewState = state.thumbnailViewState.copy(
        minWidth = minWidth,
        maxWidth = maxWidth,
        minHeight = minHeight,
        maxHeight = maxHeight
      )
    )
    state.thumbnailViewState.applyState(thumbnail)
  }

  companion object {
    private const val STATE_ROOT = "state.root"
    private const val STATE_STATE = "state.state"
  }
}
